File: 01-Programming-Languages/Java/Data-Structures/Heap/KthLargest.java
/*
 * Problem: Find the kth largest element in an array.
 * Approach: Use a min-heap of size k (PriorityQueue).
 * Time Complexity: O(n log k)
 * Space Complexity: O(k)
 * Example: Input: nums = [3,2,1,5,6,4], k = 2 -> Output: 5
 */

import java.util.*;

public class KthLargest {
    public static int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) minHeap.poll();
        }
        return minHeap.peek();
    }

    public static void main(String[] args) {
        int[] nums = {3, 2, 1, 5, 6, 4};
        int k = 2;
        System.out.println(findKthLargest(nums, k));
    }
}

File: 01-Programming-Languages/Kotlin/Data-Structures/Heap/kth_largest.kt
/*
 * Problem: Find the kth largest element in an array.
 * Approach: Use a min-heap of size k (PriorityQueue).
 * Time Complexity: O(n log k)
 * Space Complexity: O(k)
 * Example: Input: nums = [3,2,1,5,6,4], k = 2 -> Output: 5
 */

import java.util.*

fun findKthLargest(nums: IntArray, k: Int): Int {
    val minHeap = PriorityQueue<Int>()
    for (num in nums) {
        minHeap.offer(num)
        if (minHeap.size > k) minHeap.poll()
    }
    return minHeap.peek()
}

fun main() {
    val nums = intArrayOf(3, 2, 1, 5, 6, 4)
    println(findKthLargest(nums, 2))
}
